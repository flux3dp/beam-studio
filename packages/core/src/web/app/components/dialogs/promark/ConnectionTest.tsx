import React, { useCallback, useEffect, useMemo, useRef, useState } from 'react';

import { Button, Modal, Progress } from 'antd';

import { getWorkarea } from '@core/app/constants/workarea-constants';
import { swiftrayClient } from '@core/helpers/api/swiftray-client';
import { generateCalibrationTaskString, loadTaskToSwiftray } from '@core/helpers/device/promark/calibration';
import deviceMaster from '@core/helpers/device-master';
import useI18n from '@core/helpers/useI18n';
import type { IDeviceInfo } from '@core/interfaces/IDevice';
import type { ILang } from '@core/interfaces/ILang';

import styles from './ConnectionTest.module.scss';

interface Props {
  device: IDeviceInfo;
  onClose: () => void;
}

const results: Array<keyof ILang['promark_connection_test']> = ['res_0', 'res_1', 'res_2', 'res_3', 'res_4'];
const worst = results.length - 1;

const ConnectionTest = ({ device, onClose }: Props): React.JSX.Element => {
  const { global: tGlobal, promark_connection_test: t } = useI18n();
  const { model } = device;
  const { width } = useMemo(() => getWorkarea(model), [model]);
  const [isTesting, setIsTesting] = useState(false);
  const [progress, setProgress] = useState(0);
  const [result, setResult] = useState(Number.NaN);
  const previewTask = useRef<string>('');
  const timerRef = useRef<NodeJS.Timeout>();

  useEffect(() => {
    previewTask.current = '';
  }, [width]);

  const uploadPreviewTask = useCallback(async () => {
    if (!previewTask.current) {
      previewTask.current = await generateCalibrationTaskString({ width });
    }

    await loadTaskToSwiftray(previewTask.current, model);
  }, [model, width]);

  useEffect(() => {
    const abortPreview = () => {
      clearInterval(timerRef.current);
      setProgress(0);
      setIsTesting(false);
    };

    swiftrayClient.on('disconnected', abortPreview);

    return () => {
      swiftrayClient.off('disconnected', abortPreview);
      clearInterval(timerRef.current);
      deviceMaster.stopFraming();
    };
  }, []);

  const togglePreview = async (completed = false) => {
    if (!isTesting) {
      await uploadPreviewTask();
      await deviceMaster.startFraming();
      setIsTesting(true);
      setResult(Number.NaN);
      setProgress(0);
      timerRef.current = setInterval(() => setProgress((prev) => Math.min(prev + 1, 100)), 1200);
    } else {
      clearInterval(timerRef.current);
      await deviceMaster.stopFraming();
      setProgress(completed ? 100 : 0);
      setIsTesting(false);
    }
  };

  const completeTest = async () => {
    clearInterval(timerRef.current);
    await togglePreview(true);

    const { disconnection } = await deviceMaster.getReport();

    if (typeof disconnection === 'number' && disconnection >= 0) {
      setResult(Math.min(disconnection, worst));
    } else {
      setResult(worst);
    }
  };

  useEffect(() => {
    if (progress === 100) {
      completeTest();
    }
  }, [progress]);

  const passed = useMemo(() => result < 2, [result]);

  return (
    <Modal centered footer={null} keyboard={false} maskClosable={false} onCancel={onClose} open title={t.title}>
      <div className={styles.container}>
        <div>{t.description}</div>
        <Progress className={styles.progress} percent={progress} />
        {Number.isNaN(result) || (
          <div>
            <span className={styles.label}>{t.health}</span>
            <span className={passed ? styles.good : styles.bad}> {t[results[result]]}</span>
            {passed ? <div>{t.healthy_description}</div> : <div className={styles.bad}>{t.unhealthy_description}</div>}
          </div>
        )}
      </div>
      <div className={styles.footer}>
        <Button onClick={onClose}>{tGlobal.cancel}</Button>
        <Button onClick={() => togglePreview()} type="primary">
          {isTesting ? t.stop : Number.isNaN(result) ? t.start : t.restart}
        </Button>
      </div>
    </Modal>
  );
};

export default ConnectionTest;
