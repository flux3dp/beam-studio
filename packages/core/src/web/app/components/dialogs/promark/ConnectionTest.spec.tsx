import React, { act } from 'react';

import { fireEvent, render, waitFor } from '@testing-library/react';

import ConnectionTest from './ConnectionTest';

jest.useFakeTimers();

jest.mock('@core/app/constants/workarea-constants', () => ({
  getWorkarea: () => ({ width: 110 }),
}));

const mockOn = jest.fn();
const mockOff = jest.fn();

jest.mock('@core/helpers/api/swiftray-client', () => ({
  swiftrayClient: {
    off: (...args) => mockOff(...args),
    on: (...args) => mockOn(...args),
  },
}));

const mockGenerateCalibrationTaskString = jest.fn().mockResolvedValue('mock-task');
const mockLoadTaskToSwiftray = jest.fn();

jest.mock('@core/helpers/device/promark/calibration', () => ({
  generateCalibrationTaskString: (...args) => mockGenerateCalibrationTaskString(...args),
  loadTaskToSwiftray: (...args) => mockLoadTaskToSwiftray(...args),
}));

const mockStartFraming = jest.fn();
const mockStopFraming = jest.fn();
const mockGetReport = jest.fn();

jest.mock('@core/helpers/device-master', () => ({
  getReport: (...args) => mockGetReport(...args),
  startFraming: (...args) => mockStartFraming(...args),
  stopFraming: (...args) => mockStopFraming(...args),
}));

const mockDevice: any = { model: 'fbm1' };
const mockOnclose = jest.fn();

describe('test ConnectionTest', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should render correctly', async () => {
    const { baseElement, getByText } = render(<ConnectionTest device={mockDevice} onClose={mockOnclose} />);

    expect(baseElement).toMatchSnapshot();

    // Good result
    mockGetReport.mockResolvedValue({ disconnection: 0 });

    const startButton = getByText('Start Test');

    expect(startButton).toBeInTheDocument();
    fireEvent.click(startButton);
    await waitFor(() => {
      expect(mockGenerateCalibrationTaskString).toHaveBeenCalledTimes(1);
      expect(mockLoadTaskToSwiftray).toHaveBeenCalledTimes(1);
      expect(mockStartFraming).toHaveBeenCalledTimes(1);
    });

    const cancelButton = getByText('Stop Test');

    expect(cancelButton).toBeInTheDocument();
    act(() => jest.advanceTimersByTime(120000));
    await waitFor(() => {
      expect(mockStopFraming).toHaveBeenCalledTimes(1);
      expect(mockGetReport).toHaveBeenCalledTimes(1);
    });
    expect(baseElement).toMatchSnapshot();

    // Bad result
    mockGetReport.mockResolvedValue({ disconnection: -1 });

    const restartButton = getByText('Restart Test');

    expect(restartButton).toBeInTheDocument();
    fireEvent.click(restartButton);
    await waitFor(() => {
      expect(mockGenerateCalibrationTaskString).toHaveBeenCalledTimes(1);
      expect(mockLoadTaskToSwiftray).toHaveBeenCalledTimes(2);
      expect(mockStartFraming).toHaveBeenCalledTimes(2);
    });
    act(() => jest.advanceTimersByTime(120000));
    await waitFor(() => {
      expect(mockStopFraming).toHaveBeenCalledTimes(2);
      expect(mockGetReport).toHaveBeenCalledTimes(2);
    });
    expect(baseElement).toMatchSnapshot();
  });
});
