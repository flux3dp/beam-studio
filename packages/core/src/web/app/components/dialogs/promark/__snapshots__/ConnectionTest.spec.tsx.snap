// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test ConnectionTest should render correctly 1`] = `
<body>
  <div />
  <div>
    <div
      class="ant-modal-root css-dev-only-do-not-override-hash"
    >
      <div
        class="ant-modal-mask"
      />
      <div
        class="ant-modal-wrap ant-modal-centered"
        tabindex="-1"
      >
        <div
          aria-labelledby="test-id"
          aria-modal="true"
          class="ant-modal css-dev-only-do-not-override-hash ant-zoom-appear ant-zoom-appear-prepare ant-zoom"
          role="dialog"
          style="width: 520px;"
        >
          <div
            style="outline: none;"
            tabindex="0"
          >
            <div
              class="ant-modal-content"
            >
              <button
                aria-label="Close"
                class="ant-modal-close"
                type="button"
              >
                <span
                  class="ant-modal-close-x"
                >
                  <span
                    aria-label="close"
                    class="anticon anticon-close ant-modal-close-icon"
                    role="img"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="close"
                      fill="currentColor"
                      fill-rule="evenodd"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M799.86 166.31c.02 0 .04.02.08.06l57.69 57.7c.04.03.05.05.06.08a.12.12 0 010 .06c0 .03-.02.05-.06.09L569.93 512l287.7 287.7c.04.04.05.06.06.09a.12.12 0 010 .07c0 .02-.02.04-.06.08l-57.7 57.69c-.03.04-.05.05-.07.06a.12.12 0 01-.07 0c-.03 0-.05-.02-.09-.06L512 569.93l-287.7 287.7c-.04.04-.06.05-.09.06a.12.12 0 01-.07 0c-.02 0-.04-.02-.08-.06l-57.69-57.7c-.04-.03-.05-.05-.06-.07a.12.12 0 010-.07c0-.03.02-.05.06-.09L454.07 512l-287.7-287.7c-.04-.04-.05-.06-.06-.09a.12.12 0 010-.07c0-.02.02-.04.06-.08l57.7-57.69c.03-.04.05-.05.07-.06a.12.12 0 01.07 0c.03 0 .05.02.09.06L512 454.07l287.7-287.7c.04-.04.06-.05.09-.06a.12.12 0 01.07 0z"
                      />
                    </svg>
                  </span>
                </span>
              </button>
              <div
                class="ant-modal-header"
              >
                <div
                  class="ant-modal-title"
                  id="test-id"
                >
                  Connection Stability Test
                </div>
              </div>
              <div
                class="ant-modal-body"
              >
                <div
                  class="container"
                >
                  <div>
                    The system will run a red light operation for about two minutes to check if the connection is stable.
                  </div>
                  <div
                    aria-valuemax="100"
                    aria-valuemin="0"
                    aria-valuenow="0"
                    class="ant-progress ant-progress-status-normal ant-progress-line ant-progress-line-align-end ant-progress-line-position-outer ant-progress-show-info ant-progress-default progress css-dev-only-do-not-override-hash"
                    role="progressbar"
                  >
                    <div
                      class="ant-progress-outer"
                      style="width: 100%;"
                    >
                      <div
                        class="ant-progress-inner"
                      >
                        <div
                          class="ant-progress-bg ant-progress-bg-outer"
                          style="width: 0%; height: 8px; --progress-percent: 0;"
                        />
                      </div>
                      <span
                        class="ant-progress-text ant-progress-text-end ant-progress-text-outer"
                        title="0%"
                      >
                        0%
                      </span>
                    </div>
                  </div>
                </div>
                <div
                  class="footer"
                >
                  <button
                    class="ant-btn css-dev-only-do-not-override-hash ant-btn-default ant-btn-color-default ant-btn-variant-outlined"
                    type="button"
                  >
                    <span>
                      Cancel
                    </span>
                  </button>
                  <button
                    class="ant-btn css-dev-only-do-not-override-hash ant-btn-primary ant-btn-color-primary ant-btn-variant-solid"
                    type="button"
                  >
                    <span>
                      Start Test
                    </span>
                  </button>
                </div>
              </div>
            </div>
          </div>
          <div
            style="width: 0px; height: 0px; overflow: hidden; outline: none;"
            tabindex="0"
          />
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`test ConnectionTest should render correctly 2`] = `
<body>
  <div />
  <div>
    <div
      class="ant-modal-root css-dev-only-do-not-override-hash"
    >
      <div
        class="ant-modal-mask"
      />
      <div
        class="ant-modal-wrap ant-modal-centered"
        tabindex="-1"
      >
        <div
          aria-labelledby="test-id"
          aria-modal="true"
          class="ant-modal css-dev-only-do-not-override-hash"
          role="dialog"
          style="width: 520px;"
        >
          <div
            style="outline: none;"
            tabindex="0"
          >
            <div
              class="ant-modal-content"
            >
              <button
                aria-label="Close"
                class="ant-modal-close"
                type="button"
              >
                <span
                  class="ant-modal-close-x"
                >
                  <span
                    aria-label="close"
                    class="anticon anticon-close ant-modal-close-icon"
                    role="img"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="close"
                      fill="currentColor"
                      fill-rule="evenodd"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M799.86 166.31c.02 0 .04.02.08.06l57.69 57.7c.04.03.05.05.06.08a.12.12 0 010 .06c0 .03-.02.05-.06.09L569.93 512l287.7 287.7c.04.04.05.06.06.09a.12.12 0 010 .07c0 .02-.02.04-.06.08l-57.7 57.69c-.03.04-.05.05-.07.06a.12.12 0 01-.07 0c-.03 0-.05-.02-.09-.06L512 569.93l-287.7 287.7c-.04.04-.06.05-.09.06a.12.12 0 01-.07 0c-.02 0-.04-.02-.08-.06l-57.69-57.7c-.04-.03-.05-.05-.06-.07a.12.12 0 010-.07c0-.03.02-.05.06-.09L454.07 512l-287.7-287.7c-.04-.04-.05-.06-.06-.09a.12.12 0 010-.07c0-.02.02-.04.06-.08l57.7-57.69c.03-.04.05-.05.07-.06a.12.12 0 01.07 0c.03 0 .05.02.09.06L512 454.07l287.7-287.7c.04-.04.06-.05.09-.06a.12.12 0 01.07 0z"
                      />
                    </svg>
                  </span>
                </span>
              </button>
              <div
                class="ant-modal-header"
              >
                <div
                  class="ant-modal-title"
                  id="test-id"
                >
                  Connection Stability Test
                </div>
              </div>
              <div
                class="ant-modal-body"
              >
                <div
                  class="container"
                >
                  <div>
                    The system will run a red light operation for about two minutes to check if the connection is stable.
                  </div>
                  <div
                    aria-valuemax="100"
                    aria-valuemin="0"
                    aria-valuenow="100"
                    class="ant-progress ant-progress-status-success ant-progress-line ant-progress-line-align-end ant-progress-line-position-outer ant-progress-show-info ant-progress-default progress css-dev-only-do-not-override-hash"
                    role="progressbar"
                  >
                    <div
                      class="ant-progress-outer"
                      style="width: 100%;"
                    >
                      <div
                        class="ant-progress-inner"
                      >
                        <div
                          class="ant-progress-bg ant-progress-bg-outer"
                          style="width: 100%; height: 8px; --progress-percent: 1;"
                        />
                      </div>
                      <span
                        class="ant-progress-text ant-progress-text-end ant-progress-text-outer"
                      >
                        <span
                          aria-label="check-circle"
                          class="anticon anticon-check-circle"
                          role="img"
                        >
                          <svg
                            aria-hidden="true"
                            data-icon="check-circle"
                            fill="currentColor"
                            focusable="false"
                            height="1em"
                            viewBox="64 64 896 896"
                            width="1em"
                          >
                            <path
                              d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm193.5 301.7l-210.6 292a31.8 31.8 0 01-51.7 0L318.5 484.9c-3.8-5.3 0-12.7 6.5-12.7h46.9c10.2 0 19.9 4.9 25.9 13.3l71.2 98.8 157.2-218c6-8.3 15.6-13.3 25.9-13.3H699c6.5 0 10.3 7.4 6.5 12.7z"
                            />
                          </svg>
                        </span>
                      </span>
                    </div>
                  </div>
                  <div>
                    <span
                      class="label"
                    >
                      Stability:
                    </span>
                    <span
                      class="good"
                    >
                       
                      Excellent
                    </span>
                    <div>
                      Stability is good, but avoid moving the computer significantly during engraving.
                    </div>
                  </div>
                </div>
                <div
                  class="footer"
                >
                  <button
                    class="ant-btn css-dev-only-do-not-override-hash ant-btn-default ant-btn-color-default ant-btn-variant-outlined"
                    type="button"
                  >
                    <span>
                      Cancel
                    </span>
                  </button>
                  <button
                    class="ant-btn css-dev-only-do-not-override-hash ant-btn-primary ant-btn-color-primary ant-btn-variant-solid"
                    type="button"
                  >
                    <span>
                      Restart Test
                    </span>
                  </button>
                </div>
              </div>
            </div>
          </div>
          <div
            style="width: 0px; height: 0px; overflow: hidden; outline: none;"
            tabindex="0"
          />
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`test ConnectionTest should render correctly 3`] = `
<body>
  <div />
  <div>
    <div
      class="ant-modal-root css-dev-only-do-not-override-hash"
    >
      <div
        class="ant-modal-mask"
      />
      <div
        class="ant-modal-wrap ant-modal-centered"
        tabindex="-1"
      >
        <div
          aria-labelledby="test-id"
          aria-modal="true"
          class="ant-modal css-dev-only-do-not-override-hash"
          role="dialog"
          style="width: 520px;"
        >
          <div
            style="outline: none;"
            tabindex="0"
          >
            <div
              class="ant-modal-content"
            >
              <button
                aria-label="Close"
                class="ant-modal-close"
                type="button"
              >
                <span
                  class="ant-modal-close-x"
                >
                  <span
                    aria-label="close"
                    class="anticon anticon-close ant-modal-close-icon"
                    role="img"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="close"
                      fill="currentColor"
                      fill-rule="evenodd"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M799.86 166.31c.02 0 .04.02.08.06l57.69 57.7c.04.03.05.05.06.08a.12.12 0 010 .06c0 .03-.02.05-.06.09L569.93 512l287.7 287.7c.04.04.05.06.06.09a.12.12 0 010 .07c0 .02-.02.04-.06.08l-57.7 57.69c-.03.04-.05.05-.07.06a.12.12 0 01-.07 0c-.03 0-.05-.02-.09-.06L512 569.93l-287.7 287.7c-.04.04-.06.05-.09.06a.12.12 0 01-.07 0c-.02 0-.04-.02-.08-.06l-57.69-57.7c-.04-.03-.05-.05-.06-.07a.12.12 0 010-.07c0-.03.02-.05.06-.09L454.07 512l-287.7-287.7c-.04-.04-.05-.06-.06-.09a.12.12 0 010-.07c0-.02.02-.04.06-.08l57.7-57.69c.03-.04.05-.05.07-.06a.12.12 0 01.07 0c.03 0 .05.02.09.06L512 454.07l287.7-287.7c.04-.04.06-.05.09-.06a.12.12 0 01.07 0z"
                      />
                    </svg>
                  </span>
                </span>
              </button>
              <div
                class="ant-modal-header"
              >
                <div
                  class="ant-modal-title"
                  id="test-id"
                >
                  Connection Stability Test
                </div>
              </div>
              <div
                class="ant-modal-body"
              >
                <div
                  class="container"
                >
                  <div>
                    The system will run a red light operation for about two minutes to check if the connection is stable.
                  </div>
                  <div
                    aria-valuemax="100"
                    aria-valuemin="0"
                    aria-valuenow="100"
                    class="ant-progress ant-progress-status-success ant-progress-line ant-progress-line-align-end ant-progress-line-position-outer ant-progress-show-info ant-progress-default progress css-dev-only-do-not-override-hash"
                    role="progressbar"
                  >
                    <div
                      class="ant-progress-outer"
                      style="width: 100%;"
                    >
                      <div
                        class="ant-progress-inner"
                      >
                        <div
                          class="ant-progress-bg ant-progress-bg-outer"
                          style="width: 100%; height: 8px; --progress-percent: 1;"
                        />
                      </div>
                      <span
                        class="ant-progress-text ant-progress-text-end ant-progress-text-outer"
                      >
                        <span
                          aria-label="check-circle"
                          class="anticon anticon-check-circle"
                          role="img"
                        >
                          <svg
                            aria-hidden="true"
                            data-icon="check-circle"
                            fill="currentColor"
                            focusable="false"
                            height="1em"
                            viewBox="64 64 896 896"
                            width="1em"
                          >
                            <path
                              d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm193.5 301.7l-210.6 292a31.8 31.8 0 01-51.7 0L318.5 484.9c-3.8-5.3 0-12.7 6.5-12.7h46.9c10.2 0 19.9 4.9 25.9 13.3l71.2 98.8 157.2-218c6-8.3 15.6-13.3 25.9-13.3H699c6.5 0 10.3 7.4 6.5 12.7z"
                            />
                          </svg>
                        </span>
                      </span>
                    </div>
                  </div>
                  <div>
                    <span
                      class="label"
                    >
                      Stability:
                    </span>
                    <span
                      class="bad"
                    >
                       
                      Very Low
                    </span>
                    <div
                      class="bad"
                    >
                      Stability is too low. It is recommended to replace the network adapter or power cable.
                    </div>
                  </div>
                </div>
                <div
                  class="footer"
                >
                  <button
                    class="ant-btn css-dev-only-do-not-override-hash ant-btn-default ant-btn-color-default ant-btn-variant-outlined"
                    type="button"
                  >
                    <span>
                      Cancel
                    </span>
                  </button>
                  <button
                    class="ant-btn css-dev-only-do-not-override-hash ant-btn-primary ant-btn-color-primary ant-btn-variant-solid"
                    type="button"
                  >
                    <span>
                      Restart Test
                    </span>
                  </button>
                </div>
              </div>
            </div>
          </div>
          <div
            style="width: 0px; height: 0px; overflow: hidden; outline: none;"
            tabindex="0"
          />
        </div>
      </div>
    </div>
  </div>
</body>
`;
